README
NAME
    Cache::Pluggable - pluggable cache interface

SYNOPSIS
      package Proj::Cache;
      use Mouse;
      extends 'Cache::Pluggable';

      with qw(Cache::Pluggable::Plugin::Namespace);

      __PACKAGE__->meta->make_immutable();

      package main;
      use strict;
      use warnings;
      use Cache::Memcached::Fast;

      my $cache = Proj::Cache->new(
        cache => Cache::Memcached::Fast->new(
            servers => [{ address => 'localhost:11211', weight => 1 }],
            namespace => 'proj:',
        ),
      );
      my $value = $cache->get({ key => $key });
      $cache->set(key => $key, value => $value, expires_in => 60);

DESCRIPTION
    You would like to wrap CPAN's Cache::XXX libraries a bit. This library
    help you to write some modules and reuse it.

METHODS
    my $cache = Proj::Cache->new(%args);
        %args is followings:

        cache: Object ( required )
            Cache::xxx object like Cache::Memcached::Fast.

    my $value = $cache->get($hash_ref);
        yet anothor get interface for extend get interface. (recommend)

        $hash_ref are followings:

        key: Str ( required )
                key name to get.

    my $value = $cache->get($key);
        many Cache:: library has this interface. But, It's hard to extend.
        So, $hash_ref is recommend.

    $cache->set($hash_ref);
        $hash_ref is followings:

        key: Str ( required )
        value: Str ( required )
        expires_in: Int ( optional )
            expires_in in N seconds.

    $cache->set($key, $value, $expire);
        many Cache:: library has this intercace. But, It's hard to extend.
        So, $hash_ref is recommend.

TESTED cache module.
    Cahce::Pluggable work with followings: +<Cahce::Memcached::Fast>

AUTHOR
    Keiji Yoshimi <walf443 at gmail dot com>

SEE ALSO
    +<CHI>

    +<Cache::Cache>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Cache::Pluggable

0.01    Wed Mar 16 21:00:19 2011
        - original version


lib/Test/Cache/Pluggable.pm
package Test::Cache::Pluggable;
use Mouse;
use Test::More;
use Test::Differences qw/eq_or_diff/;
use Data::Dumper;
use Proc::Guard qw();
use Test::TCP qw();
use File::Which qw();
use Try::Tiny;

has 'cache' => (
    is => 'ro',
);

no Mouse;

sub lives_ok {
    my ($cb, $msg) = @_;

    my $error;
    try { $cb->() } catch {
        $error = $_;
    };
    ok(! defined $error, $msg)
        or diag($error);
}

sub guard_service {
    my ($class, $cmd, @args) = @_;

    my $program = File::Which::which('memcached')
        or plan skip_all => q{This test require 'memcached'};

    my $port = Test::TCP::empty_port();
    my $proc = Proc::Guard::proc_guard($program, map { $_ eq '$port' ? $port : $_ } @args);
    Test::TCP::wait_port($port);

    return ($proc, $port);
}

sub guard_memcached {
    my ($class, ) = @_;
    return $class->guard_service('memcached', '-l', '127.0.0.1', '-p', '$port');
}

sub run {
    my ($self, $hashref, $test_name) = @_;

    $test_name ||= "Test::Cache::Pluggable#run";

    subtest $test_name => sub {
        subtest 'set/get testing' => sub {
            subtest 'hashref interface' => sub {
                lives_ok(sub {
                    $self->cache->set($hashref);
                }, 'set ok');

                my $value;
                lives_ok(sub {
                    $value = $self->cache->get($hashref);
                }, 'get ok');

                eq_or_diff($value, $hashref->{value}, "get/set ok")
                    or Dumper($value);

            };

            subtest 'key interface' => sub {
                lives_ok(sub {
                    my $val = $self->cache->set($hashref->{key} => $hashref->{value});
                }, 'set ok');

                my $value;
                lives_ok(sub {
                    $value = $self->cache->get($hashref->{key});
                }, 'get ok');

                eq_or_diff($value, $hashref->{value}, "get/set ok")
                    or Dumper($value);

            };
        };
    };

}

1;
__END__

=head1 NAME

Test::Cache::Pluggable - test for your plugin

=head1 SYNOPSIS

    package t::Cache;
    use Mouse;
    extends 'Cache::Pluggable';
    with qw(YourPlugin);

    package main;
    use Test::Cache::Pluggable;

    my ($guard, $port) = Test::Cache::Pluggable->guard_memcached();

    my $cache = t::Cache->new(
        cache => $cache
    );
    my $c = Test::Cache::Pluggable->new($cache);
    $c->run({ key => "foo", value => "bar", extends => 10 });

=head1 DESCRIPTION

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<CLI>

+<Cache::Cache>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Cache/Pluggable/Plugin/Duplicate.pm
package Cache::Pluggable::Plugin::Duplicate;
use Mouse::Role;

has 'duplicate' => (
    is => 'ro',
    isa => "ArrayRef",
    default => sub {
        [],
    },
);

around '_set' => sub {
    my $orig = shift;
    my $self = shift;

    my @args = @_;
    $orig->($self, @args);
    for my $cache ( @{ $self->duplicate } ) {
        $cache->set(@args);
    }
};

no Mouse::Role;

1;

__END__
=head1 NAME

Cache::Pluggable::Plugin::Duplicate - duplicate data

=head1 SYNOPSIS

  package Proj::Cache;
  use Mouse;
  extends 'Cache::Pluggable';

  with qw(Cache::Pluggable::Plugin::Duplicate);

  __PACKAGE__->meta->make_immutable();

  package main;
  use strict;
  use warnings;
  use Cache::Memcached::Fast;

  my $migrate_memcache = Cache::Memcached::Fast->new(
        servers => [{ address => 'host2:11211', weight => 1 }],
  );
  my $cache = Proj::Cache->new(
    cache => Cache::Memcached::Fast->new(
        servers => [{ address => 'host1:11211', weight => 1 }],
    ),
    duplicate => [
        $migrate_memcache,
    ],
  );
  $cache->set("hoge", "fuga"); # set hoge to host1 and host2.

=head1 DESCRIPTION

This plugin duplicate set data. It's useful to migrate cache storage.

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<Cache::Pluggable>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Cache/Pluggable/Plugin/Namespace.pm
package Cache::Pluggable::Plugin::Namespace;
use Mouse::Role;

has 'namespace' => (
    is  => 'rw',
    isa => 'Str',
);

around 'key_filter' => sub {
    my $orig = shift;
    my $self = shift;
    my $result = $orig->($self, @_);
    $self->namespace . ':' . $result;
};

no Mouse::Role;

1;

__END__

=head1 NAME

Cache::Pluggable::Plugin::Namespace - added namespace to key.

=head1 SYNOPSIS

  package Proj::Cache;
  use Mouse;
  extends 'Cache::Pluggable';

  with qw(Cache::Pluggable::Plugin::Namespace);

  __PACKAGE__->meta->make_immutable();

  package main;
  use strict;
  use warnings;
  use Cache::Memcached::Fast;

  my $cache = Proj::Cache->new(
    namespace => "proj",
    cache => Cache::Memcached::Fast->new(
        servers => [{ address => 'localhost:11211', weight => 1 }],
    ),
  );
  my $value = $cache->get($key); # "GET proj:$key"
  $cache->set(foo => $value, $expire); # "SET proj:$key $value"

=head1 DESCRIPTION

This plugin always append namespace to key.

It's useful to share cache storages between many projects.

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<Cache::Pluggable>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Cache/Pluggable/Plugin/Serialize.pm
package  Cache::Pluggable::Plugin::Serialize;
use Mouse::Role;

has 'serialize_methods' => (
    is => 'ro',
    isa => 'ArrayRef[CodeRef]',
    default => sub { [
        sub { JSON::XS->new->utf8->encode(shift) },
        sub { JSON::XS->new->utf8->decode(shift) },
    ] },
);

around '_get' => sub {
    my $orig = shift;
    my $self = shift;
    my $ret = $orig->($self, @_);
    return $self->serialize_methods->[1]->($ret);
};

around '_set_hashref' => sub {
    my $orig = shift;
    my $self = shift;
    my $hashref = shift;
    my %hash = %$hashref; # copy for privending side effect.
    $hash{value} = $self->serialize_methods->[0]->($hash{value});
    $orig->($self, \%hash, @_);
};

no Mouse::Role;

1;

__END__
=head1 NAME

Cache::Pluggable::Role::Namespace - added namespace to key.

=head1 SYNOPSIS

  package Proj::Cache;
  use Mouse;
  extends 'Cache::Pluggable';

  with qw(Cache::Pluggable::Plugin::Serialize);

  __PACKAGE__->meta->make_immutable();

  package main;
  use strict;
  use warnings;
  use Cache::Memcached::Fast;

  my $cache = Proj::Cache->new(
    cache => Cache::Memcached::Fast->new(
        servers => [{ address => 'localhost:11211', weight => 1 }],
    ),
    serialize_methods => [ sub { JSON::XS->new->utf8->encode(shift) }, sub { JSON::XS->new->utf8->decode(shift) } ], # it's default serializer.
  );
  my $value = $cache->get($key); # "GET proj:$key"
  $cache->set(foo => $value, $expire); # "SET proj:$key $value"

=head1 DESCRIPTION

This plugin for serializing data to storage.
Cache::Memcached::Fast support serializing data by Storable as default.
But some Cache library don't support serialization.

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<Cache::Pluggable>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Cache/Pluggable/Plugin/GetCallback.pm
package Cache::Pluggable::Plugin::GetCallback;
use Mouse::Role;
no Mouse::Role;

sub get_callback {
    my ($self, $hash_ref, $code_ref) = @_;
    my $value = $self->_get_hashref($hash_ref);
    return $value if defined $value;
    return if $hash_ref->{not_callback};
    $value = $code_ref->($hash_ref->{key});
    $self->set({ 
        %$hash_ref,
        value => $value || undef,
    });
    return $value;
}

1;

__END__
=head1 NAME

Cache::Pluggable::Plugin::GetCallback - add get_callback interface

=head1 SYNOPSIS

  package Proj::Cache;
  use Mouse;
  extends 'Cache::Pluggable';

  with qw(Cache::Pluggable::Plugin::GetCallback);

  __PACKAGE__->meta->make_immutable();

  package main;
  use strict;
  use warnings;
  use Cache::Memcached::Fast;

  my $cache = Proj::Cache->new(
    cache => Cache::Memcached::Fast->new(
        servers => [{ address => 'localhost:11211', weight => 1 }],
    ),
  );
  my $value = $cache->get_callback({
     key => $key,
     expires_in => 10, # sec
  } => sub {
      # do heavy task.
      my $value = $key x 3;
      return $value;
  });

=head1 DESCRIPTION

This plugin add get_callback method.

=over

=item my $value = $self->get_callback($hash_ref => $code_ref);

$hash_ref is followings:

=over 4

=item key: Str

key to get and set.

=item expires_in: Int ( optional )

expire in N sec.

=item not_callback: Bool ( optional )

If it was set, don't run $code_ref.
It may be useful in situation when you want to call $code_ref from CLI
and you does not want to call $code_ref from Web.

=back

$code_ref is CodeRef. It should return set value.

=back

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<Cache::Pluggable>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Cache/Pluggable.pm
package Cache::Pluggable;
use Mouse;
use Class::Load qw();
our $VERSION = '0.01';

has 'cache' => (
    is => 'rw',
);

__PACKAGE__->meta->make_immutable;
no Mouse;

sub get {
    my $self = shift;
    if ( ref $_[0] eq 'HASH' ) {
        return $self->_get_hashref(@_);
    } else {
        my $key = shift;
        return $self->_get_hashref({ key => $key }, @_);
    }
}

sub key_filter {
    my (undef, $key) = @_;
    return $key;
}

sub _get_hashref {
    my $self = shift;
    my $hash_ref = shift;
    my %hash = %$hash_ref; # copy
    return $self->_get($self->key_filter($hash{key}), @_);
}

sub _get {
    my $self = shift;
    $self->cache->get(@_);
}

sub set {
    my $self = shift;
    if ( ref $_[0] eq 'HASH' ) {
        return $self->_set_hashref(@_);
    } else {
        my $key = shift;
        my $value = shift;
        my $expires_in = shift;
        return $self->_set_hashref({
            key => $key || undef,
            value => $value || undef,
            expires_in => $expires_in || undef,
        }, @_);
    }
}

sub _set_hashref {
    my $self = shift;
    my $hash_ref = shift;
    my %hash = %$hash_ref; # copy
    return $self->_set($self->key_filter($hash{key}), $hash{value}, $hash{expires_in} || undef, @_);
}

sub _set {
    my $self = shift;
    $self->cache->set(@_);
}

1;
__END__

=head1 NAME

Cache::Pluggable - pluggable cache interface

=head1 SYNOPSIS

  package Proj::Cache;
  use Mouse;
  extends 'Cache::Pluggable';

  with qw(Cache::Pluggable::Plugin::Namespace);

  __PACKAGE__->meta->make_immutable();

  package main;
  use strict;
  use warnings;
  use Cache::Memcached::Fast;

  my $cache = Proj::Cache->new(
    cache => Cache::Memcached::Fast->new(
        servers => [{ address => 'localhost:11211', weight => 1 }],
        namespace => 'proj:',
    ),
  );
  my $value = $cache->get({ key => $key });
  $cache->set(key => $key, value => $value, expires_in => 60);

=head1 DESCRIPTION

You would like to wrap CPAN's Cache::XXX libraries a bit.
This library help you to write some modules and reuse it.

=head1 METHODS

=over

=item my $cache = Proj::Cache->new(%args);

%args is followings:

=over 4

=item cache: Object ( required )

Cache::xxx object like Cache::Memcached::Fast.

=back

=item my $value = $cache->get($hash_ref);

yet anothor get interface for extend get interface. (recommend)

$hash_ref are followings:

=over 4

=item key: Str ( required )

    key name to get.

=back

=item my $value = $cache->get($key);

many Cache:: library has this interface. 
But, It's hard to extend. So, $hash_ref is recommend.

=item $cache->set($hash_ref);

$hash_ref is followings:

=over 4

=item key: Str ( required )

=item value: Str ( required )

=item expires_in: Int ( optional )

expires_in in N seconds.

=back

=item $cache->set($key, $value, $expire);

many Cache:: library has this intercace.
But, It's hard to extend. So, $hash_ref is recommend.

=back

=head1 TESTED cache module.

Cahce::Pluggable work with followings:
+<Cahce::Memcached::Fast>

=head1 AUTHOR

Keiji Yoshimi E<lt>walf443 at gmail dot comE<gt>

=head1 SEE ALSO

+<CHI>

+<Cache::Cache>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
